protocol/serialization-types/src/common.rs

```diff
@@ -16,15 +17,77 @@ pub type Hash2V1 = Versioned<HashV1, 1>;
 
 /// u64 representing a token ID (v1)
 pub type TokenIdV1 = Versioned<Versioned<Versioned<u64, 1>, 1>, 1>;
+impl_from_for_newtype!(U64Json, TokenIdV1);
 
 /// u64 representing a block time (v1)
 pub type BlockTimeV1 = Versioned<Versioned<u64, 1>, 1>;
 
 /// u64 representing an account nonce (v1) // This should also be an extendedu32
 pub type AccountNonceV1 = Versioned<Versioned<u32, 1>, 1>;
 
+/// u32 wrapper (json)
+/// Note that integers are represented as string in mina json
+#[derive(Clone, Debug, PartialEq, derive_more::From)]
+pub struct U32Json(pub u32);
+
+impl Serialize for U32Json {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let s = format!("{}", self.0);
+        serializer.serialize_str(&s)
+    }
+}
+
+impl<'de> Deserialize<'de> for U32Json {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let s = String::deserialize(deserializer)?;
+
+        Ok(Self(
+            s.parse().map_err(<D::Error as serde::de::Error>::custom)?,
+        ))
+    }
+}
+
+/// u64 wrapper (json)
+/// Note that integers are represented as string in mina json
+#[derive(Clone, Debug, PartialEq, derive_more::From)]
+pub struct U64Json(pub u64);
+
+impl Serialize for U64Json {
```

Comment:
I wonder if this can be shortened with: https://serde.rs/variant-attrs.html#serialize_with

Reply:
Unfortunately, this only saves 1 line or 2 but it requires another mod and it's not straightforward for ppl who read the code to understand what is going on unless they are very familiar with serde attributes